glommio/src/executor/stall.rs

```diff
@@ -0,0 +1,283 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the MIT/Apache-2.0 License, at your convenience
+//
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2021 Datadog, Inc.
+//
+#![allow(dead_code)]
+
+use nix::sys;
+use std::{
+    sync::{
+        atomic::{AtomicBool, Ordering},
+        Arc,
+    },
+    thread::JoinHandle,
+    time::{Duration, Instant},
+};
+
+#[cfg(feature = "stall-detection")]
+thread_local! {
+    static STALL_STRACE_COLLECTOR: (
+        crossbeam::channel::Sender<backtrace::BacktraceFrame>,
+        crossbeam::channel::Receiver<backtrace::BacktraceFrame>
+    ) = crossbeam::channel::bounded(1 << 10);
+}
+
+#[cfg(feature = "stall-detection")]
+lazy_static! {
+    static ref STALL_SIG_HANDLER: signal_hook::SigId = unsafe {
+        signal_hook::low_level::register(signal_hook::consts::SIGUSR1, move || {
+            STALL_STRACE_COLLECTOR.with(|collector| {
+                if collector.0.is_full() {
+                    return;
+                }
+
+                backtrace::trace_unsynchronized(|frame| {
+                    collector
+                        .0
+                        .try_send(backtrace::BacktraceFrame::from(frame.clone()))
+                        .is_ok()
+                });
+            });
+        })
+    }
+    .unwrap();
+}
+
+#[derive(Debug)]
+pub(crate) struct StallDetector {
+    timer: Arc<sys::timerfd::TimerFd>,
+    handler: Option<JoinHandle<()>>,
+    id: usize,
+    terminated: Arc<AtomicBool>,
+}
+
+impl StallDetector {
+    #[cfg(feature = "stall-detection")]
+    pub(crate) fn new(executor_id: usize) -> std::io::Result<StallDetector> {
+        lazy_static::initialize(&STALL_SIG_HANDLER);
+
+        let timer = Arc::new(
+            sys::timerfd::TimerFd::new(
+                sys::timerfd::ClockId::CLOCK_MONOTONIC,
+                sys::timerfd::TimerFlags::empty(),
+            )
+            .map_err(std::io::Error::from)?,
+        );
+        let tid = unsafe { nix::libc::pthread_self() };
+        let terminated = Arc::new(AtomicBool::new(false));
+        let timer_handler = std::thread::spawn(enclose::enclose! { (terminated, timer) move || {
+            while timer.wait().is_ok() {
+                if terminated.load(Ordering::Acquire) {
+                    return
+                }
+                unsafe { nix::libc::pthread_kill(tid, nix::libc::SIGUSR1) };
+            }
+        }});
+
+        Ok(Self {
+            timer,
+            handler: Some(timer_handler),
+            id: executor_id,
+            terminated,
+        })
+    }
+
+    pub(crate) fn enter_task_queue(
+        &self,
+        queue_name: String,
+        start: Instant,
+        max_expected_runtime: Duration,
+    ) -> nix::Result<StallDetectorGuard<'_>> {
+        // We consider a queue to be stalling the system if it failed to yield in due
```

Comment:
no, this should be configurable by the user when they configure the stall detector in the specific reactor.
Let's say you do know of a stall that happens sometimes, but don't prioritize fixing it now: at least your logs shouldn't be spammed by it